interop: Interop Activation Handling

tyler-smith opened the pull request:
.

protolambda reviewed (commented):
We'll need the op-node to act on resets pre-interop, since the supervisor cannot resolve the reset for the op-node without indexed data.

tcrypt25519 committed:
tweak,interop: Change CanInitiateAt/CanExecuteAt checks to > instead of >=.

ghost commented:
Corresponding spec change PR for the change to the CanInitiateAt check to be `>` instead of `>=`: https://github.com/ethereum-optimism/specs/pull/691

ajsutton commented on op-supervisor/supervisor/backend/db/update.go:
```diff
@@ -50,6 +57,20 @@ func (db *ChainsDB) initializedSealBlock(chain eth.ChainID, block eth.BlockRef)
 }
 
+func (db *ChainsDB) isInteropActive(timestamp uint64) bool {
+	if db.depSet == nil {
+		return true
+	}
```
I don't have context so could easily be wrong, but this seems unsafe - shortly if we don't have a dependency set interop would be inactive?

tyler-smith commented on op-supervisor/supervisor/backend/activation/activation.go:
```diff
@@ -0,0 +1,208 @@
+		return false
+	}
+
+	for _, chain := range am.depSet.Chains() {
```
Bug: We need handle activation chain-by-chain.

axelKingsley replied:
Additional reasoning here:

- Application: if chains get added, their later activation would make the supervisor appear as deactivated until that new chain aged into interop.
- Protocol: if some chains have activated, but other chains haven't, those active chains *must* be able to do interop replacements.

I think this means that activation needs to be a per-chain concept

tyler-smith commented on op-supervisor/supervisor/backend/backend.go:
```diff
@@ -176,10 +196,52 @@ func NewSupervisorBackend(ctx context.Context, logger log.Logger,
+		// If this is a block in an interop period but we we don't have an initialized database,
+		// we need to get an anchor block and initialize
+		isInitialized := su.chainDBs.IsInitialized(x.ChainID)
+		if !isInitialized && isBlockInInteropPeriod {
```
The check on isBlockInInteropPeriod is duplicated

axelKingsley commented on op-supervisor/supervisor/backend/backend.go:
```diff
@@ -176,10 +196,52 @@ func NewSupervisorBackend(ctx context.Context, logger log.Logger,
 func (su *SupervisorBackend) OnEvent(ev event.Event) bool {
 	switch x := ev.(type) {
 	case superevents.LocalUnsafeReceivedEvent:
+		// Check if the block is in an interop active period for this chain
```
Let's check that unsafe blocks are a good trigger for protocol upgrades. Maybe safe is the better trigger, since it is *safe*.

tyler-smith replied:
I think we landed on using Finalized blocks.

ajsutton replied:
I'm not quite sure what it means to use finalized as the trigger here. If interop only finalizes when a block with the interop timestamp finalises then the fault proof system will be broken because there's no record on chain of when blocks become safe or finalized.  In the fault proof system the agreed block from the bisection process is the finalized, safe and unsafe head (because both sides agreed to it so it must be canonical and will never change).

It's ok for the supervisor to tell the sequencer to not accept executing messages until interop has finalized, but the derivation rules for interop, including both accepting valid executing messages and re-orging out invalid ones MUST be applied in every block with a timestamp >= the interop timestamp. It probably works out fine if op-supervisor waits for the first interop block to be finalized to start indexing so long as it uses that block to start its index and correctly evaluates all the existing blocks so it doesn't re-org out blocks created prior to things finalizing that contain valid executing messages just because supervisor hasn't yet indexed the initiating messages.

Unresolving this so the fault proof implications don't get lost.

tyler-smith commented on op-supervisor/supervisor/backend/activation/activation.go:
```diff
@@ -0,0 +1,208 @@
+	anchorTime := time.Unix(int64(anchor.Derived.Time), 0)
+	window := time.Duration(am.depSet.MessageExpiryWindow()) * time.Second
+
+	if time.Since(anchorTime) > window {
```
Bug: This can't be right, otherwise you couldn't sync after `expiry window` seconds after interop activates.

protolambda commented on op-e2e/actions/helpers/l2_verifier.go:
```diff
@@ -294,6 +294,10 @@ func (s *l2VerifierBackend) SetRecoverMode(ctx context.Context, mode bool) error
 }
 
+func (s *l2VerifierBackend) ResetToL1(ctx context.Context, l1BlockNumber uint64) error {
+	return nil
```
This should not silently no-op.

protolambda commented on op-node/node/api.go:
```diff
@@ -100,6 +101,12 @@ func (n *adminAPI) SetRecoverMode(ctx context.Context, mode bool) error {
 
+// RequestReset requests a reset of the derivation pipeline to start from a specific L1 block
+// This is used to handle reset scenarios, including resetting to pre-interop state
+func (n *adminAPI) RequestReset(ctx context.Context, l1BlockNumber uint64) error {
```
This is used by the Supervisor, and so should not be attached to the admin namespace but to the interop namespace in the managed-mode package.

protolambda commented on op-node/rollup/driver/state.go:
```diff
@@ -507,6 +507,40 @@ func (s *Driver) SetRecoverMode(ctx context.Context, mode bool) error {
+	// Emit event to reset the engine, send a block event, and trigger processing
+	s.emitter.Emit(engine.ResetEngineRequestEvent{})
+	s.emitter.Emit(status.L1UnsafeEvent{L1Unsafe: l1Block})
+	s.emitter.Emit(StepReqEvent{ResetBackoff: true})
```
This makes no sense to me, what does it do?

protolambda commented on op-supervisor/supervisor/backend/activation/activation.go:
```diff
@@ -0,0 +1,204 @@
+	GetAnchorBlock(chainID eth.ChainID) (eth.BlockRef, error)
+}
+
+type ActivationManager struct {
```
This needs docs, it's hard to follow what it does

protolambda commented on op-supervisor/supervisor/backend/depset/mock.go:
```diff
@@ -0,0 +1,68 @@
+	"github.com/ethereum-optimism/optimism/op-supervisor/supervisor/types"
+)
+
+type MockDependencySet struct {
```
Why mock a dependency set when a regular static dependency set is so simple to configure? When do we need these callbacks?

protolambda commented on op-supervisor/supervisor/backend/syncnode/reset.go:
```diff
@@ -10,10 +10,6 @@ import (
 	"github.com/ethereum/go-ethereum"
 )
 
-// resetTracker manages a bisection
-// between consistent and inconsistent blocks
-// and is used to prepare a reset request
-// which is sent to the managed node
```
Why remove all the comments?

tyler-smith replied:
Fixed.

ajsutton commented on op-supervisor/supervisor/backend/backend.go:
```diff
@@ -173,29 +203,201 @@ func NewSupervisorBackend(ctx context.Context, logger log.Logger,
+		su.logger.Info("Interop activation detected, need to get anchor point",
+			"chain", chainID, "block", block)
+
+		// In tests, there may not be sync sources available
+		syncSrc, ok := su.syncSources.Get(chainID)
+		if !ok || syncSrc == nil {
+			// In tests we'll just create a mock anchor point
+			if testing.Testing() {
+				// For tests, just initialize with a mock anchor point
+				mockAnchor := types.DerivedBlockRefPair{
+					Source:  eth.BlockRef{Number: 0, Hash: common.HexToHash("0x123"), Time: 1},
+					Derived: eth.BlockRef{Number: 0, Hash: common.HexToHash("0x456"), Time: 1},
+				}
+				su.chainDBs.InitializeWithAnchor(chainID, mockAnchor)
+				su.logger.Info("Test environment: Initialized with mock anchor point", "chain", chainID)
+				return nil
+			}
+
+			// In production, log and continue
+			su.logger.Error("No sync source available for chain", "chain", chainID)
+			return fmt.Errorf("no sync source available for chain %s", chainID)
+		}
```
Just flagging that having test specific code in the middle of key production logic is a huge smell.  Why is it that sync sources can't be made available in tests but will be 100% guaranteed to be available in production? Shouldn't tests be going through the same genesis and setup process as new chains would?

ghost commented:
I've pushed a new version that:

- Addresses all review comments except which activation block to use (Unsafe, safe, or finalized)
- Reduces the amount of code added
- Reduces the amount of code changed
- Makes new RequestReset RPC endpoint use engine reset event
- Adds better testing
- Updates CanExecuteAt semantics to match new CanInitiateAt

It has remaining issues:

- The reset logic isn't working correctly in tests and maybe in brand new chains. We try to issue a `ResetEngineRequestEvent` but that fails because the node has no finalized blocks yet.
- Afaict for things to work correctly we want blocks without executing messages to be considered cross-valid before interop activates; but we determine that by checking our DBs and we aren't maintaining those pre-interop.

protolambda commented on op-supervisor/supervisor/backend/depset/static.go:
```diff
@@ -124,7 +124,7 @@ func (ds *StaticConfigDependencySet) CanInitiateAt(chainID eth.ChainID, initTime
 		return false, nil
 	}
-	return initTimestamp >= dep.HistoryMinTime, nil
+	return initTimestamp > dep.HistoryMinTime, nil
```
This will work for now, but the HistoryMinTime timestamp doesn't always align with the activation block timestamp. So this can be wrong in some configurations.

sebastianst replied:
Just for my understanding, what's the background on this change?

protolambda reviewed (approved):
Looks ok, but we'll need to fix the timestamp edge-case, and handle better how we go about finding a stable anchor block that gets finalized, in case the chain reorgs shortly after upgrade, to a different anchor-block.

tcrypt25519 committed:
feature: Add interop activation check helpers.

tcrypt25519 committed:
feature: Add interop_blockRefByLabel RPC method.

tcrypt25519 committed:
feature: Handle interop activation.

tcrypt25519 committed:
tests: Increase checks in interop upgrade acceptance test.

tcrypt25519 committed:
feature: Add pre-interop head tracking.

tcrypt25519 committed:
Fix lint issues.

tcrypt25519 committed:
fix: On AnchorEvent, exit pre activation mode.

tcrypt25519 committed:
refactor, cleanup

sebastianst commented on op-supervisor/supervisor/backend/activation/activation.go:
```diff
@@ -0,0 +1,38 @@
+}
+
+// Check checks if interop is active for a given chain and timestamp.
+func (c *Check) Check(chain eth.ChainID, timestamp uint64) bool {
```
Couldn't this just be a function on the `DependencySet` struct, with return type `(bool, error)` and then let the caller decide how to handle an error (e.g., log)?

sebastianst commented on op-supervisor/supervisor/backend/db/db.go:
```diff
@@ -147,11 +159,31 @@ func NewChainsDB(l log.Logger, depSet depset.DependencySet, m Metrics) *ChainsDB
+
+	// Pre-initialize chains in pre-activation mode if appropriate
+	for _, chainID := range depSet.Chains() {
+		canActivate, err := depSet.CanInitiateAt(chainID, uint64(time.Now().Unix()))
+		if err != nil {
+			l.Warn("Failed to check activation status for chain", "chain", chainID, "err", err)
+			continue
+		}
```
Does it make sense to continue here if the check fails? Shouldn't we return the error instead?

sebastianst commented on op-supervisor/supervisor/backend/db/preactivation.go:
```diff
@@ -0,0 +1,221 @@
+
+// ExitPreActivationMode transitions a chain from pre-activation to normal mode
+// This should be called when interop activation is detected
+func (db *ChainsDB) ExitPreActivationMode(id eth.ChainID, anchor types.DerivedBlockRefPair) error {
```
maybe rename to something like `ActivateInterop`?

sebastianst commented on op-supervisor/supervisor/backend/db/anchor.go:
```diff
@@ -14,14 +14,18 @@ func (db *ChainsDB) ForceInitialized(id eth.ChainID) {
+	db.initFromAnchor(id, anchor)
+}
+
 func (db *ChainsDB) initFromAnchor(id eth.ChainID, anchor types.DerivedBlockRefPair) {
```
Just for my understanding, why don't we let `initFromAnchor` return an `error` and instead just warn-log if initialization fails? It seems quite critical to me if this call were to fail, shouldn't the caller be made aware of it and act accordingly (possible failing the whole application with a critical error)?

sebastianst commented on op-supervisor/supervisor/backend/db/preactivation.go:
```diff
@@ -0,0 +1,221 @@
+			"current", nodeStatus.LocalUnsafe.Number,
+			"new", block.Number)
+		return
+	}
```
Why this? An unsafe reorg may rewind to an earlier block, and then the old unsafe block with higher block id would be invalid.

sebastianst commented on op-supervisor/supervisor/backend/db/preactivation.go:
```diff
@@ -0,0 +1,221 @@
+
+	// Only update LocalSafe if the new block is newer
+	currentLocalSafe := nodeStatus.LocalSafe
+	if derivedBlockSeal.Number > currentLocalSafe.Number {
```
same question, what about reorgs?

sebastianst commented on op-supervisor/supervisor/backend/db/preactivation.go:
```diff
@@ -0,0 +1,221 @@
+			Number:     derivedBlockSeal.Number,
+			Time:       derivedBlockSeal.Timestamp,
+			ParentHash: common.Hash{}, // We don't have parent hash in BlockSeal
+		}
```
Could we create a helper for this, like `BlockRefFromSeal`? Also wondering, isn't it a problem that the parent hash is missing? Are we sure there's no code hitting this block ref that assumes that a parent is set?

sebastianst commented on op-supervisor/supervisor/backend/db/query.go:
```diff
@@ -10,6 +10,31 @@ import (
+		}
+
+		// Return the head as a block seal
+		return types.BlockSeal{
+			Hash:      head.Hash,
+			Number:    head.Number,
+			Timestamp: head.Time,
+		}, nil
```
we should create a helper in `types` like `HeaderBlockSeal` that returns a `BlockSeal` for a header, I guess this expression comes up a lot.

sebastianst commented on op-supervisor/supervisor/backend/db/query.go:
```diff
@@ -10,6 +10,31 @@ import (
+			if number > head.Number {
+				return types.BlockSeal{}, types.ErrFuture
+			}
+			return types.BlockSeal{}, types.ErrConflict
```
So it's guaranteed that all calls to the functions in this file, in pre-activation mode, aren't called for older blocks, not at tip?

sebastianst commented on op-supervisor/supervisor/backend/db/query.go:
I'm generally wondering, in a pre-activation op-supervisor, why are any of these functions even called on the `ChainsDB`? Shouldn't the higher-level logic just not do most of its operations if the timestamps in question are pre-Interop?

sebastianst commented on op-supervisor/supervisor/backend/syncnode/controller.go:
```diff
@@ -29,17 +31,20 @@ type SyncNodesController struct {
 	return &SyncNodesController{
-		logger:   l,
-		depSet:   depset,
-		eventSys: eventSys,
-		backend:  backend,
+		logger:          l,
+		depSet:          depset,
+		eventSys:        eventSys,
+		backend:         backend,
+		activationCheck: activationCheck,
```
Why do we pass the `activationCheck` here? Can't it just be recreated from the `depSet` and `logger`? Note also that I suggest we remove this type and just add the check function to the `DependencySet` struct.

sebastianst commented on op-supervisor/supervisor/backend/syncnode/controller.go:
```diff
@@ -81,7 +86,7 @@ func (snc *SyncNodesController) AttachNodeController(chainID eth.ChainID, ctrl S
 
 	// create the managed node, register and return
-	node := NewManagedNode(logger, chainID, ctrl, snc.backend, noSubscribe)
+	node := NewManagedNode(logger, chainID, ctrl, snc.backend, snc.activationCheck, noSubscribe)
```
Is this the reason for the `ActivationCheck` type? If yes, I think it's better to let the `ManagedMode` consume an interface that has the `Check...` function that it needs, and then just pass the dep set here.

sebastianst commented on op-supervisor/supervisor/backend/syncnode/node.go:
```diff
@@ -207,6 +214,24 @@ func (m *ManagedNode) WatchSubscriptionErrors() {
 
 func (m *ManagedNode) Start() {
+	// Initialize database in pre-activation mode if needed
+	if !m.activationCheck.Check(m.chainID, uint64(time.Now().Unix())) {
```
Why does the current system time matter? Shouldn't it be the L2 block time?

sebastianst commented on op-supervisor/supervisor/backend/syncnode/rpc.go:
```diff
@@ -69,6 +69,21 @@ func (rs *RPCSyncNode) BlockRefByNumber(ctx context.Context, number uint64) (eth
 }
 
+func (rs *RPCSyncNode) BlockRefByLabel(ctx context.Context, label eth.BlockLabel) (eth.BlockRef, error) {
+	var out *eth.BlockRef
+	err := rs.cl.CallContext(ctx, &out, "interop_blockRefByLabel", label)
+	if err != nil {
+		var jsonErr gethrpc.Error
+		if errors.As(err, &jsonErr) {
+			if jsonErr.ErrorCode() == 0 { // TODO
+				return eth.BlockRef{}, ethereum.NotFound
+			}
+		}
+		return eth.BlockRef{}, err
+	}
+	return *out, nil
```
```suggestion
	var out eth.BlockRef
	err := rs.cl.CallContext(ctx, &out, "interop_blockRefByLabel", label)
	if err != nil {
		var jsonErr gethrpc.Error
		if errors.As(err, &jsonErr) {
			if jsonErr.ErrorCode() == 0 { // TODO
				return eth.BlockRef{}, ethereum.NotFound
			}
		}
		return eth.BlockRef{}, err
	}
	return out, nil
```

sebastianst replied:
What about the TODO? I see it in other places in this file too.

sebastianst reviewed (commented):
Part I - still need to review some tests.

tcrypt25519 committed:
fix: Preserve cross-unsafe values during pre-activation to normal mode transition

- Simplify approach by making any anchor event exit pre-activation mode
- Add explicit cross-unsafe updates during local-unsafe updates
- Preserve cross-unsafe values when exiting pre-activation mode
- Fix type consistency between BlockSeal and BlockRef in preactivation methods
- Remove debugging print statements and simplify logging
